### DISCUSSION TITLE
Repeated closures of a stream can lead to program freezing, which is undesirable behavior.

### DISCUSSION THREAD

**Discussion by rootBiao** [strongswan/strongswan#2204] [Q&A] [answered]
![image](https://github.com/strongswan/strongswan/assets/87700793/fd321e86-a1cd-4996-b5a2-be244fe9df73)
![image](https://github.com/strongswan/strongswan/assets/87700793/b40e4dae-079a-49ca-abec-7f1587a594f5)

When a stream is closed for the second time, the program halts instead of proceeding further. Ideally, upon attempting to close an already closed stream, the program should promptly return, avoiding any stalls.

**Comment by rootBiao** [comment]
![image](https://github.com/strongswan/strongswan/assets/87700793/08ea48da-8a79-43b8-956b-adf3a491d49f)

![image](https://github.com/strongswan/strongswan/assets/87700793/67248ee4-e207-46b6-a3c0-1ec0ad97ec20)
Even if called once, it will be blocked.

**Comment by tobiasbrunner** [comment]
You can't destroy an object more than once. After calling `destroy()` you must not access that pointer anymore.

**Reply by rootBiao** [reply]
![image](https://github.com/strongswan/strongswan/assets/87700793/20b8e5d4-2788-41d2-9044-e4abeecb1886)

![image](https://github.com/strongswan/strongswan/assets/87700793/738a8a26-1753-4ed3-89eb-b3133f3bb9cc)

Even if called once, it will be blocked.

**Reply by tobiasbrunner** [reply]
What do you mean? And obviously check your own code first.

**Reply by rootBiao** [reply]
![image](https://github.com/strongswan/strongswan/assets/87700793/b2d3f4a7-e529-4728-8ad3-d5d876261a8b)

![image](https://github.com/strongswan/strongswan/assets/87700793/fdbc18e3-ec84-4ba7-88d7-315ab9a63996)

![image](https://github.com/strongswan/strongswan/assets/87700793/8751be1d-e2f9-4c1e-aa88-95948aaba808)
![image](https://github.com/strongswan/strongswan/assets/87700793/1af2e453-da7f-4f71-b639-02d0407ce823)
Does it mean that there is no need to actively execute the destroy function to release resources when the return value of the on_ppe_read function is false?

**Reply by tobiasbrunner** [reply, marked as answer]
You do have to destroy it if you are done with it. But you currently can't do it directly from the `on_read` callback of the stream because the `watcher_t::remove()` method called by `destroy()` waits for all callbacks to return, which obviously causes a deadlock if called from such a callback.  This is actually documented: https://github.com/strongswan/strongswan/blob/3a2017032451aecb2921d19684f8fde8e8893428/src/libstrongswan/networking/streams/stream.h#L40-L41

So you have to do this asynchronously, similar to this (`destroy_entry()` destroys the stream here): https://github.com/strongswan/strongswan/blob/3a2017032451aecb2921d19684f8fde8e8893428/src/libcharon/plugins/vici/vici_socket.c#L284-L314

**Reply by rootBiao** [reply]
Thank you！！！
